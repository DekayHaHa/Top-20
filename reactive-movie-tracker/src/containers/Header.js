import React, { Component } from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import { signInUser, addFavorites } from "../actions/index";
import PropTypes from "prop-types";
import "../styles/Header.scss";

export const Header = class extends Component {
  clearUserData = () => {
    this.props.signInUser(0, "")
    this.props.addFavorites([]);
  };

  render() {
    const { activeUser } = this.props
    const toggleBtn = activeUser.id ? true : false
    const userTitle = `Welcome Back ${activeUser.name}!`
    const noUserTitle = 'Movie Tracker'
    return (
      <div className="Header">
        <Link to="/login"><button className="btn" disabled={toggleBtn}>User Sign In</button></Link>
        <h2 className="title">{activeUser.id ? userTitle : noUserTitle}</h2>
        <Link to="/login">
          <button className="btn" disabled={!toggleBtn} onClick={this.clearUserData}>Sign Out?</button>
        </Link>
      </div>
    );
  }
};

Header.propTypes = {
  activeUser: PropTypes.object
};

export const mapStateToProps = state => ({
  activeUser: state.activeUser
});

export const mapDispatchtoProps = dispatch => ({
  signInUser: (id, name) => dispatch(signInUser(id, name)),
  addFavorites: movies => dispatch(addFavorites(movies))
});

export default connect(
  mapStateToProps,
  mapDispatchtoProps
)(Header);