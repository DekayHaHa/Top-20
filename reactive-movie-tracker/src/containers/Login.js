import React, { Component } from "react";
import { connect } from "react-redux";
import { Redirect } from "react-router-dom";
import PropTypes from "prop-types";
import "../styles/Login.scss";
import {createUser} from '../Thunks/createUser'
import { signIn } from '../Thunks/signIn'

export class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: "",
      password: "",
      email: ""
    };
  }

  newUser = e => {
    e.preventDefault();    
    const { createUser } = this.props
    const userInfo = {
      id: 1,
      name: this.state.name,
      password: this.state.password,
      email: this.state.email
    };
    createUser(userInfo)
  };

  handleSignIn = e => {
    e.preventDefault();
    const { signIn } = this.props
    const userInfo = {
      name: this.state.name,
      password: this.state.password,
      email: this.state.email
    };
    signIn(userInfo)
  };

  handleNameInput = event => {
    this.setState({
      name: event.target.value
    });
  };

  handlePasswordInput = event => {
    this.setState({
      password: event.target.value
    });
  };

  handleEmailInput = event => {
    this.setState({
      email: event.target.value
    });
  };

  signInInputs = () => {
    const { email, password } = this.state;
    return email && password ? false : true;
  };

  signUpInputs = () => {
    const { email, password, name } = this.state;
    return email && password && name ? false : true;
  };

  render() {
    const { email, password, name } = this.state;
    let formSubmitMethod = this.handleSignIn;
    const { error } = this.props
    if (error) {
      formSubmitMethod = this.handlePost;
    }
    const signInBtnToggle = this.signInInputs();
    const signUpBtnToggle = this.signUpInputs();
    return (
      <div className="Login">
        {this.props.activeUser.id > 0 && <Redirect to="/" />}
        {error && <h3 className="wrong-password">{error}</h3>}
        <form className="form" onSubmit={formSubmitMethod}>
          {error && (
            <label className="label">
              <input
                className='name-input'
                placeholder='Full Name'
                type="text"
                value={name}
                onChange={this.handleNameInput} />
            </label>
          )}
          <label label className="label">    
            <input
              className='email-input'
              placeholder='Email Address'
              type="text"
              value={email}
              onChange={this.handleEmailInput} />
          </label>
          <label label className="label">
            <input
              className='password-input'
              placeholder='Password'
              type="password"
              value={password}
              onChange={this.handlePasswordInput}
            />
          </label>
          <button className='btn signIn-btn' disabled={signInBtnToggle} onClick={this.handleSignIn}>
            Sign In
          </button>
          {error && (
            <button className='btn signUp-btn' disabled={signUpBtnToggle} onClick={this.newUser}>
              Sign Up NOW!
            </button>
          )}
        </form>
      </div>
    );
  }
}

Login.propTypes = {
  activeUser: PropTypes.object
};

export const mapStateToProps = state => ({
  activeUser: state.activeUser,
  error: state.error
});

export const mapDispatchtoProps = dispatch => ({
  createUser: (userInfo) => dispatch(createUser(userInfo)),
  signIn: (userInfo) => dispatch(signIn(userInfo)),
});

export default connect(
  mapStateToProps,
  mapDispatchtoProps
)(Login);
