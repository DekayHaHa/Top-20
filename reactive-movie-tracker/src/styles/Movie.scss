.movie {
	width: 33%;
	border: 1px solid black;
	margin: 20px 0px;
	position: relative;
	display: flex;
	flex-direction: column;
	justify-content: space-between;
	padding: 10px;
}

.movie-title {
	text-align: center;
	margin: 0px;
}

img {
	width: 100%;
}

.favorite-btn {
	// width: 65px;
	// height: 65px;
	position: absolute;
	top: 14px;
	left: 10px;
	background: none;
	border: none;
	// background-image: url('./iconmonstr-star-6.svg')

	&:hover {
    background: lightseagreen;
  }
}