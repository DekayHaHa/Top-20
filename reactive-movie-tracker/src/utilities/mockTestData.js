export const mockReduxMovies = [
	{	title: 'Kung Fu',
	image: 'www.reddit.com',
	overview: 'Big Blue House',
	score: 7.8,
	id: 166428,
	isFavorite: false,
	releaseDate: "2019-01-03",},
	{	title: 'Kung Fu',
	image: 'www.reddit.com',
	overview: 'Big Blue House',
	score: 7.8,
	id: 166428,
	isFavorite: false,
	releaseDate: "2015-01-03",},
	{	title: 'Spider Man',
	image: 'www.reddit.com',
	overview: 'Big Blue House',
	score: 9,
	id: 123455,
	isFavorite: false,
	releaseDate: "2017-01-03",},
]

export const mockReduxUser = {
	name: 'Jeff',
	id: 2,
}

export const mockReduxFavs = [
	{title: 'Spider Man',
	image: 'www.reddit.com',
	overview: 'Big Blue House',
	score: 9,
	id: 123455,
	isFavorite: true,
	releaseDate: "2017-01-03",}
]