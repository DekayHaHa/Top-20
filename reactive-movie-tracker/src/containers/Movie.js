import React, { Component } from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import { addFavorites, addMovies } from "../actions";
import { retrieveAllFavorites } from "../utilities/api";
import PropTypes from "prop-types";
import "../styles/Movie.scss";

export const Movie = class extends Component {
  addToFavorites = async () => {
    const {
      id,
      activeUser,
      title,
      image,
      releaseDate,
      score,
      overview
    } = this.props;

    const url = "http://localhost:3000/api/users/favorites/new";
    const movieInfo = {
      movie_id: id,
      user_id: activeUser.id,
      title: title,
      poster_path: image,
      release_date: releaseDate,
      vote_average: score,
      overview: overview
    };
    await fetch(url, {
      method: "POST",
      body: JSON.stringify(movieInfo),
      headers: {
        "Content-Type": "application/json"
      }
    });
    activeUser.id && (await this.retrieveAllFavFromAPI());
  };
  retrieveAllFavFromAPI = async () => {
    const { activeUser } = this.props;
    const id = activeUser.id;
    const url = `http://localhost:3000/api/users/${id}/favorites`;
    let results = await retrieveAllFavorites(url);
    await this.compareFavorites(results);
    await this.props.addFavorites(results);
  };

  deleteFromFavorites = async () => {
    const { id, activeUser } = this.props;

    const user_id = activeUser.id;
    const movie_id = id;
    const url = `http://localhost:3000/api/users/${user_id}/favorites/${movie_id}`;
    const movieInfo = {
      movie_id: id,
      user_id: activeUser.id
    };
    await fetch(url, {
      method: "DELETE",
      body: JSON.stringify(movieInfo),
      headers: {
        "Content-Type": "application/json"
      }
    });
    activeUser.id && (await this.retrieveAllFavFromAPI());
  };

  compareFavorites = favorites => {
    const { movies } = this.props;
    const favIds = favorites.map(fav => fav.id);
    const newMovies = movies.map(movie => {
      return favIds.includes(movie.id)
        ? { ...movie, isFavorite: true }
        : { ...movie, isFavorite: false };
    });
    this.props.addMovies(newMovies);
  };

  componentDidMount() {
    this.props.activeUser.id && this.retrieveAllFavFromAPI();
  }

  render() {
    const { id, title, image, isFavorite, activeUser } = this.props;
    let methodToggle;
    let btnVal;
    if (isFavorite) {
      methodToggle = this.deleteFromFavorites;
      btnVal = "Unfavorite";
    } else {
      methodToggle = this.addToFavorites;
      btnVal = "Favorite";
    }
    const favBtn = (
      <button className={`favorite ${isFavorite}`} onClick={methodToggle}>
        {btnVal}
      </button>
    );
    const redirectBtn = (
      <Link to="/login">
        <button className={`favorite ${isFavorite}`}>Favorite</button>
      </Link>
    );
    const movieButton = activeUser.id ? favBtn : redirectBtn;
    return (
      <div>
        <Link to={`/movie/${id}`}>
          <h1>{title}</h1>
          <img alt="movie poster" src={image} />
        </Link>
        {movieButton}
      </div>
    );
  }
};

Movie.propTypes = {
  movies: PropTypes.array,
  activeUser: PropTypes.object,
  favorites: PropTypes.array
};

export const mapStateToProps = state => ({
  movies: state.movies,
  activeUser: state.activeUser,
  favorites: state.favorites
});

export const mapDispatchToProps = dispatch => ({
  addFavorites: movies => dispatch(addFavorites(movies)),
  addMovies: movies => dispatch(addMovies(movies))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Movie);
