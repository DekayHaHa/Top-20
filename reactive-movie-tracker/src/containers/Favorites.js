import React, { Component } from "react";
import { cleanFavorites} from '../utilities/cleaner'
import Movie from "./Movie";
import { connect } from "react-redux";
import { retrieveAllFavorites} from '../utilities/api'
import { addMovies, addFavorites } from "../actions/index"


class Favorites extends Component {


}

export const mapStateToProps = state => ({
  movies: state.movies,
  activeUser: state.activeUser,
  favorites: state.favorites
});

export const mapDispatchToProps = dispatch => ({
  addMovies: (movies) => dispatch(addMovies(movies)),
  addFavorites: (movie) => dispatch(addFavorites(movie))
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Favorites);
