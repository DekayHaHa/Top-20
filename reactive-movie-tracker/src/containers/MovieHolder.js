import React, { Component } from "react";
import { cleanFavorites} from '../utilities/cleaner'
import Movie from "./Movie";
import { connect } from "react-redux";
import { addMovies } from "../actions/index"

class MovieHolder extends Component {
  constructor(props) {
    super(props);
    this.state = {
      displayFavorites: false,
      favorites: []
    };
  }

  retrieveAllFavorites = async () => {
    const id = this.props.activeUser.id;
    const url = `http://localhost:3000/api/users/${id}/favorites`;
    const response = await fetch(url, {
      method: "GET",
      headers: {
        "Content-Type": "application/json"
      }
    });
		const data = await response.json();
    const favorites = await cleanFavorites(data.data)
		await this.setState({favorites}, () => this.compareFavorites())
   
  };

  toggleFavorites = () => {
    this.setState({ displayFavorites: this.state.displayFavorites ? false : true})
  };

  compareFavorites = () => {
    // return movies compared against user favorites
    const { movies } = this.props
    const { favorites } = this.state
    const favIds = favorites.map(fav => fav.id)
    const newMovies = movies.map(movie => {
      return favIds.includes(movie.id) ? {...movie, isFavorite: true} : movie
    })
    this.props.addMovies(newMovies)
  }

  render() {
    const { movies, activeUser } = this.props
    const { displayFavorites, favorites } = this.state
    const btnText = displayFavorites ? 'Show All' : 'Display Favorites';
    const moviesToRender = activeUser.id && displayFavorites ? favorites : movies
    {activeUser.id && this.retrieveAllFavorites()}
      return (
        <div>
          <button onClick={this.toggleFavorites}>{btnText}</button>
          {moviesToRender.map(movie => <Movie key={movie.id} {...movie} />)}
        </div>
      );
    // } else {
		// 	// console.log(this.state.favorites)
    //   return (
    //     <div>
    //       <button onClick={this.toggleFavorites}>Display All Movies</button>
    //       <div>
    //         {this.state.favorites.map(movie => {
    //           return (
    //             <div key={movie.id}>
    //               <Movie {...movie} />
    //             </div>
    //           );
    //         })}
    //       </div>
    //     </div>
    //   );
    
  }
}

export const mapStateToProps = state => ({
  movies: state.movies,
  activeUser: state.activeUser
});

export const mapDispatchToProps = dispatch => ({
  addMovies: (movies) => dispatch(addMovies(movies))
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MovieHolder);
