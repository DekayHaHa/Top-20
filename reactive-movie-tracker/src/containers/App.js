import React, { Component } from "react";
import MovieHolder from "./MovieHolder";
import { fetchData } from "../utilities/api";
import { APIkey } from "../utilities/key.js";
import { connect } from "react-redux";
import { addMovies } from "../actions";
import Header from "./Header";
import { Route, Switch } from "react-router-dom";
import Login from "./Login";
import { cleanMovies } from "../utilities/cleaner";
import { MovieDetails } from "../components/MovieDetails";
import { PageNotFound } from "../components/PageNotFound";
import PropTypes from "prop-types";
import "../styles/App.scss";

export class App extends Component {
  componentDidMount() {
    this.getMovieData();
  }
  getMovieData = async () => {
    const url = `https://api.themoviedb.org/3/movie/popular?api_key=${APIkey}&language=en-US&page=1`;
    try {
      //isloading
      const response = await fetchData(url);
      const finalData = cleanMovies(response.results);
      this.props.addMovies(finalData);
    } catch(error) {
      //finish out
    }
  };
  findMovieToRender = id => {
    const MovieToRender = this.props.movies.find(movie => movie.id === id);
    const movie = <MovieDetails {...MovieToRender} />;
    return MovieToRender ? movie : false;
  };
  render() {
    return (
      <div className="App">
        <Route path="/" component={Header} />
        <Switch>
          <Route exact path="/" component={MovieHolder} />
          <Route exact path="/login" component={Login} />
          <Route
            exact
            path="/movie/:id"
            render={({ match }) => {
              const movie = this.findMovieToRender(parseInt(match.params.id));
              return movie || <PageNotFound />;
            }}
          />
          <Route component={PageNotFound} />
        </Switch>
      </div>
    );
  }
}

App.propTypes = {
  movies: PropTypes.array,
  activeUser: PropTypes.object
};

export const mapStateToProps = state => ({
  movies: state.movies,
  activeUser: state.activeUser
});

export const mapDispatchtoProps = dispatch => ({
  addMovies: movies => dispatch(addMovies(movies))
});

export default connect(
  mapStateToProps,
  mapDispatchtoProps
)(App);
