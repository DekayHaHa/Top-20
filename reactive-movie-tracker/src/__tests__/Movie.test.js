import React from 'react';
import { shallow } from "enzyme";
import { Movie, mapDispatchToProps, mapStateToProps } from '../containers/Movie'
import { mockReduxFavs, mockReduxMovies, mockReduxUser, mockReduxNewMovies} from '../utilities/mockTestData'


const mockFunc = jest.fn()
describe("Movie", () => {
	let wrapper;
	beforeEach(() => {
		wrapper = shallow(<Movie 
			addMovies={mockFunc}
			movies={mockReduxMovies}
			activeUser={mockReduxUser}
			favorites={mockReduxFavs}
		/>);
	});

	it("should match the snapshot with all data passed in correctly", () => {
		expect(wrapper).toMatchSnapshot();
	});

	it.skip("should add movie to favorites", () => {
		//setup
		//execution
		//expectation
	});
	it.skip("should retireve all favorites", () => {
		//setup
		//execution
		//expectation
	});
	it.skip("should delete a movie from favorites", () => {
		//setup
		//execution
		//expectation
	});
	it("should compare favorites and change movies data", () => {
		//setup
		//execution
		wrapper.instance().compareFavorites(mockReduxFavs)
		//expectation
		expect(wrapper.instance().props.addMovies).toBeCalledWith(mockReduxNewMovies)
	});
	it.skip("should map state to props", () => {
		//setup
		//execution
		//expectation
	});
	it.skip("should map dispatch to props", () => {
		//setup
		//execution
		//expectation
	});
});