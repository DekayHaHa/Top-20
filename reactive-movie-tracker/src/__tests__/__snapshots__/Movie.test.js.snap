// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie should match the snapshot with all data passed in correctly 1`] = `
<div>
  <Link
    to="/movie/undefined"
  >
    <h1 />
    <img
      alt="movie poster"
    />
  </Link>
  <button
    className="favorite undefined"
    onClick={[Function]}
  >
    Favorite
  </button>
</div>
`;
